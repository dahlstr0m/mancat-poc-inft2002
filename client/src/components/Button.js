// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

/**
 * Renders a success button using Bootstrap styles.
 */
class ButtonSuccess extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-success' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a danger button using Bootstrap styles.
 */
class ButtonDanger extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-danger' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a light button using Bootstrap styles.
 */
class ButtonLight extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-light' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a button using Bootstrap styles.
 */
export default class Button {
  static Success = ButtonSuccess;
  static Danger = ButtonDanger;
  static Light = ButtonLight;
}
