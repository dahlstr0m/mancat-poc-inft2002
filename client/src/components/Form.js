// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

/**
 * Renders a form label using Bootstrap styles.
 */
class FormLabel extends Component<{ children?: React.Node }> {
  render() {
    return <label className="col-form-label">{this.props.children}</label>;
  }
}

/**
 * Renders a form input using Bootstrap styles.
 */
class FormInput extends Component<{
  type: string,
  value: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders a form textarea using Bootstrap styles.
 */
class FormTextarea extends React.Component<{
  value: React.Node,
  onChange: (SyntheticEvent<HTMLTextAreaElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, rows, cols
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, ...rest } = this.props;
    return <textarea {...rest} className="form-control" value={value} onChange={onChange} />;
  }
}

/**
 * Renders a form checkbox using Bootstrap styles.
 */
class FormCheckbox extends Component<{
  checked: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
  labeling: React.Node,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { labeling, checked, onChange, ...rest } = this.props;
    return (
      <div className="custom-control custom-checkbox">
        <input
          {...rest}
          type="checkbox"
          className="custom-control-input"
          id="customCheck"
          checked={checked}
          onChange={onChange}
        />
        <label className="custom-control-label" htmlFor="customCheck">
          {labeling}
        </label>
      </div>
    );
  }
}

/**
 * Renders a form select using Bootstrap styles.
 */
class FormSelect extends Component<{
  value: React.Node,
  onChange: (SyntheticEvent<HTMLSelectElement>) => mixed,
  children?: React.Node,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, size.
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, children, ...rest } = this.props;
    return (
      <select {...rest} className="custom-select" value={value} onChange={onChange}>
        {children}
      </select>
    );
  }
}

/**
 * Renders a form date using Bootstrap styles.
 */
class FormDate extends Component<{
  type: string,
  value: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders form components using Bootstrap styles.
 */
export default class Form {
  static Label = FormLabel;
  static Input = FormInput;
  static Textarea = FormTextarea;
  static Checkbox = FormCheckbox;
  static Select = FormSelect;
  static Date = FormDate;
}
