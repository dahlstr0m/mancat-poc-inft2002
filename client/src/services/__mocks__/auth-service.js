// @flow

type User = {
  username: string,
  password: string,
};

class AuthService {
  authenticated: boolean;

  constructor() {
    this.authenticated = false;
  }

  login(user: User) {
    if (user.username === 'admin' && user.password === 'admin') {
      this.authenticated = true;
      return Promise.resolve();
    }
  }

  isAuthenticated() {
    return this.authenticated;
  }
}
const authService = new AuthService();
export default authService;
