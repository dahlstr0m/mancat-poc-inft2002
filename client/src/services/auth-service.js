// @flow

import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v1';

export type User = {
  username: string,
  password: string,
};

class AuthService {
  authenticated: boolean;

  constructor() {
    this.authenticated = false;
  }

  login(user: User) {
    return axios
      .post<User, {}>('/auth/login', user)
      .then(() => (this.authenticated = true))
      .catch((error) => console.log(error));
  }

  isAuthenticated() {
    return this.authenticated;
  }
}

const authService = new AuthService();
export default authService;
