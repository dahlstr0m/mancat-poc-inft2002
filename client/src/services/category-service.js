// @flow
import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v1';

export type Category = {
  categoryId: number,
  categoryName: string,
};

class CategoryService {
  getCategories() {
    return axios.get<Category[]>('/categories').then((response) => response.data);
  }

  getCategory(categoryId: number) {
    return axios.get<Category>('/categories/' + categoryId).then((response) => response.data);
  }

  getProjectCategory(projectId: number) {
    return axios.get<Category>(`/projects/${projectId}/category`).then((response) => response.data);
  }

  createCategory(category: Category) {
    return axios
      .post<Category, { categoryId: number }>('/categories', category)
      .then((response) => response.data);
  }

  updateCategory(category: Category) {
    return axios
      .put<Category, void>(`/categories/${category.categoryId}`, category)
      .then((response) => response.data);
  }

  deleteCategory(categoryId: number) {
    return axios.delete<void>('/categories/' + categoryId).then((response) => response.data);
  }
}

const categoryService = new CategoryService();
export default categoryService;
