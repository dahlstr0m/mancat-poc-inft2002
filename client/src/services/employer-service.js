// @flow
import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v1';

export type Employer = {
  employerId: number,
  employerName: string,
};

class EmployerService {
  getEmployers() {
    return axios.get<Employer[]>('/employers').then((response) => response.data);
  }

  getEmployer(employerId: number) {
    return axios.get<Employer>('/employers/' + employerId).then((response) => response.data);
  }

  getProjectEmployer(projectId: number) {
    return axios.get<Employer>(`/projects/${projectId}/employer`).then((response) => response.data);
  }

  createEmployer(employer: Employer) {
    return axios
      .post<Employer, { employerId: number }>('/employers', employer)
      .then((response) => response.data);
  }

  updateEmployer(employer: Employer) {
    return axios
      .put<Employer, void>(`/employers/${employer.employerId}`, employer)
      .then((response) => response.data);
  }

  deleteEmployer(employerId: number) {
    return axios.delete<void>('/employers/' + employerId).then((response) => response.data);
  }
}

const employerService = new EmployerService();
export default employerService;
