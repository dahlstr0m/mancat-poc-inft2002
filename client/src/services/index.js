export * from './auth-service';
export * from './portfolio-service';
