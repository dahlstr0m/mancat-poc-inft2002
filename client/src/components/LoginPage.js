import * as React from 'react';
import { Component } from 'react-simplified';
import Form from './Form';
import Button from './Button';
import { Card } from './Card';
import { Alert } from './Widgets';
import authService from '../services/auth-service';
import { history } from '../index';

/**
 * Renders a simple page with username and password fields.
 */
export default class LoginPage extends Component {
  user = { username: '', password: '' };

  render() {
    return (
      <Card title="Log in">
        <Form.Label>Username:</Form.Label>
        <Form.Input
          type="text"
          value={this.user.username}
          onChange={(event) => (this.user.username = event.currentTarget.value)}
        />
        <Form.Label>Password:</Form.Label>
        <Form.Input
          type="password"
          value={this.user.password}
          onChange={(event) => (this.user.password = event.currentTarget.value)}
        />
        <hr />
        {/* Takes the user to the admin page if the login was successful. */}
        <Button.Success
          onClick={() =>
            authService.login(this.user).then(() => {
              if (authService.isAuthenticated()) {
                history.push('/admin');
              }
            })
          }
        >
          Log in
        </Button.Success>
      </Card>
    );
  }
}
