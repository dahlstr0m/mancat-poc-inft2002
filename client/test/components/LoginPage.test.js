// @flow

import * as React from 'react';
import { shallow } from 'enzyme';

// Types
import { type User } from '../../src/services/auth-service';

// Components
import Form from '../../src/components/Form';
import Button from '../../src/components/Button';
import LoginPage from '../../src/components/LoginPage';

// Mocks
jest.mock('../../src/services/auth-service');

// Tests
describe('LoginPage component tests', () => {
  test('LoginPage draws correctly', (done) => {
    const wrapper = shallow(<LoginPage />);

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();

      done();
    });
  });

  test('Login redirects correctly when authenticated', (done) => {
    const wrapper = shallow(<LoginPage />);

    wrapper
      .find(Form.Input)
      .at(0)
      .simulate('change', { currentTarget: { value: 'admin' } });
    wrapper
      .find(Form.Input)
      .at(1)
      .simulate('change', { currentTarget: { value: 'admin' } });

    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Input type="text" value="admin" />)).toEqual(true);
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Input type="password" value="admin" />)).toEqual(
      true
    );

    wrapper.find(Button.Success).simulate('click');

    setTimeout(() => {
      expect(location.hash).toEqual('#/admin');

      done();
    });
  });
});
