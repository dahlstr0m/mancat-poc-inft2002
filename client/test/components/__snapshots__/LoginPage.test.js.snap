// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage component tests LoginPage draws correctly 1`] = `
<Card
  title="Log in"
>
  <FormLabel>
    Username:
  </FormLabel>
  <FormInput
    onChange={[Function]}
    type="text"
    value=""
  />
  <FormLabel>
    Password:
  </FormLabel>
  <FormInput
    onChange={[Function]}
    type="password"
    value=""
  />
  <hr />
  <ButtonSuccess
    onClick={[Function]}
  >
    Log in
  </ButtonSuccess>
</Card>
`;
